server/cluster/unsafe_recovery_controller.go

```diff
@@ -120,462 +184,747 @@ func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]st
+				u.changeStage(createEmptyRegion)
+			}
+		case forceLeader, forceLeaderForCommitMerge:
+			if u.generateForceLeaderPlan(newestRegionTree, peersMap, true) {
+				u.changeStage(forceLeaderForCommitMerge)
+			} else if u.generateForceLeaderPlan(newestRegionTree, peersMap, false) {
+				u.changeStage(forceLeader)
+			} else if u.generateDemoteFailedVoterPlan(newestRegionTree, peersMap) {
+				u.changeStage(demoteFailedVoter)
+			} else if u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap) {
+				u.changeStage(createEmptyRegion)
 			}
-		} else if report, exist := u.storeReports[heartbeat.Stats.StoreId]; exist && report == nil {
-			u.storeReports[heartbeat.Stats.StoreId] = heartbeat.StoreReport
+		case demoteFailedVoter:
+			// may still have plan to do, recheck again
+			if u.generateForceLeaderPlan(newestRegionTree, peersMap, false) {
+				u.changeStage(forceLeader)
+			} else if u.generateDemoteFailedVoterPlan(newestRegionTree, peersMap) {
+				u.changeStage(demoteFailedVoter)
+			} else if u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap) {
+				u.changeStage(createEmptyRegion)
+			}
+		case createEmptyRegion:
+			if u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap) {
+				u.changeStage(createEmptyRegion)
+			}
+		default:
+			panic("unreachable")
+		}
+
+		hasPlan := len(u.storeRecoveryPlans) != 0
+		if u.err != nil {
+			u.changeStage(failed)
+			return
+		} else if !hasPlan {
+			u.changeStage(finished)
+			return
+		}
+	}
+
+	u.dispatchPlan(heartbeat, resp)
+}
+
+/// It dispatches recovery plan if any.
+func (u *unsafeRecoveryController) dispatchPlan(heartbeat *pdpb.StoreHeartbeatRequest, resp *pdpb.StoreHeartbeatResponse) {
+	storeID := heartbeat.Stats.StoreId
+	now := time.Now()
+
+	if reported, exist := u.storeReports[storeID]; reported != nil || !exist {
+		// the plan has been executed, no need to dispatch again
+		// or no need to displan plan to this store(e.g. Tiflash)
+		return
+	}
+
+	if expire, dispatched := u.storePlanExpires[storeID]; !dispatched || expire.Before(now) {
+		if dispatched {
+			log.Info("Unsafe recovery store recovery plan execution timeout, retry", zap.Uint64("store-id", storeID))
+		}
+		// Dispatch the recovery plan to the store, and the plan may be empty.
+		resp.RecoveryPlan = u.getRecoveryPlan(storeID)
+		resp.RecoveryPlan.Step = u.step
+		u.storePlanExpires[storeID] = now.Add(storeRequestInterval)
+	}
+}
+
+// It collects and checks if store reports have been fully collected.
+func (u *unsafeRecoveryController) collectReport(heartbeat *pdpb.StoreHeartbeatRequest) (bool, error) {
+	storeID := heartbeat.Stats.StoreId
+	if _, isFailedStore := u.failedStores[storeID]; isFailedStore {
+		return false, errors.Errorf("Receive heartbeat from failed store %d", storeID)
+	}
+
+	if heartbeat.StoreReport == nil {
+		return false, nil
+	}
+
+	if heartbeat.StoreReport.GetStep() != u.step {
+		log.Info("Unsafe recovery receives invalid store report",
```

Comment:
Does the log level need to be changed?

Reply:
it's not an error, it's possible when suffering network issues